src/adapter/src/coord/message_handler.rs

```diff
@@ -676,4 +697,105 @@ impl<S: Append + 'static> Coordinator<S> {
             });
         }
     }
+
+    #[tracing::instrument(level = "debug", skip_all)]
+    /// Finishes sequencing a command that was waiting on a real time recency timestamp.
+    async fn message_real_time_recency_timestamp(
+        &mut self,
+        conn_id: ConnectionId,
+        transient_revision: u64,
+        (compute_instance_id, compute_instance_name): (ComputeInstanceId, String),
+        id_bundle: CollectionIdBundle,
+        mut object_names: HashMap<GlobalId, String>,
+        real_time_recency_ts: mz_repr::Timestamp,
+    ) {
+        if let Some(real_time_recency_context) =
+            self.pending_real_time_recency_timestamp.remove(&conn_id)
+        {
+            // Re-validate that the involved objects still exist.
+            if transient_revision != self.catalog.transient_revision() {
```

Comment:
This logic *looks* correct to me but makes some assumptions: items never mutate (probably safe?), and that compute instance ids and catalog ids are the full set of things that need to be checked when the revision doesn't match. This may or may not be true today, let's assume it is. In the future, the actual checking logic that exists in `sequence_peek_begin` could change and we might forget to update or check this location.

Is it possible that if the transient revision has changed, we re-call sequence_peek_begin so that it can do full checking again? This would make me feel much better about the future safety of this code.

Reply:
I think an issue with that is that `sequence_peek_begin` doesn't do enough validations and makes some assumptions about the catalog that may have changed. As an alternative, maybe we should carry the original command along with us, and if the revision has changed start over all the way at the command level.

Reply:
Taking a closer look, re-executing the original command won't work because of read then writes. There's no easy way to know if the current peak was for a regular peak or for a read then write. Therefore we don't know the original command that we'll need to resend. I'll try and find out a way around this.

Reply:
The more I look at this the more difficult I think it will be. I think that for the purposes of this PR we'll be overly cautious and fail if the transient revision has changed no matter what. Since this feature is still gated behind unsafe mode and concurrent DDL is going to be a fairly rare occurrence, I think it's fine to leave this as future work. I'll open a ticket.

Reply:
This sounds fine for now. Can we preserve the original statement and re-plan that from the top?

Reply:
(For the later PR, not here.)

Reply:
I think we'll have to do something along those lines. The real complication comes from being able to differentiate between read then writes and plain reads.